File: Week3/Code/Vectorize1.py
"""Comparison of element-wise matrix sum to built-in vectorized sum function"""

__appname__ = 'Vectorize1.py'
__author__ = 'Sam Turner (sat19@ic.ac.uk)'
__version__ = '0.0.1'
__license__ = 'GNU public' 

# imports
import numpy as np
import time

# random 1000 * 1000 matrix
M = np.random.random([1000,1000])

def sum_all_elements(M):
    """
    Calculate matrix sum by elementwise addition
    PARAMETERS
    ----------
    M : np.array
        array to sum
    
    RETURNS
    -------
    int
        sum of elements of matrix
    """
    dims  = M.shape
    t=0
    for i in range(dims[0]):
        for j in range(dims[1]):
            t += M[i,j]
    return t


# time elementwise function
start = time.time()
sum_all_elements(M)
end = time.time()

sum_elems_time = end-start

# time builtin funtion

start = time.time()
np.sum(M)
end = time.time()

sum_np_time = end-start


# print results

print(" Python\n--------")

print("Unvectorized:\t" + "{0:.3f}".format(sum_elems_time))

print("Vect (NumPy):\t" + "{0:.3f}".format(sum_np_time))

File: Week3/Code/Vectorize1.R


## Script:  Vectorize1.R
## Author:  Sam Turner sat19@ic.ac.uk
## About:   Comparison of element-wise matrix sum to built in vectorized sum function

# generate 1000 * 1000 matrix of random numbers
M <- matrix(runif(1000000),1000,1000)

# sum all elements of matrix by looping
SumAllElements <- function(M){
  Dimensions <- dim(M)
  Tot <- 0
  for (i in 1:Dimensions[1]){
    for (j in 1:Dimensions[2]){
      Tot <- Tot + M[i,j]
    }
  }
  return (Tot)
}
 
# Time built-in and loopy methods of matrix summing
cat(" R \n---\n")

cat("Unvectorized:\t", system.time(SumAllElements(M))[[3]],"\n")

cat("Vectorized:\t", system.time(sum(M))[[3]],"\n\n")
